### PR TITLE
Separate Sapling provers

### PR DESCRIPTION
**Pull request by str4d**
Part of #752 (maybe closes it, to be determined by what #741 requires).
Conflicts with #999 (whichever is merged later will be rebased).

### CODE REVIEW

**Comment by str4d** [comment]
[Rebased on `main`](https://github.com/zcash/librustzcash/compare/73b6cc006a3e5f9bc74513bf39379467a758e32b..4d17970c51079a323327cf1ae240846e9c05d6e9) to make changes after #999 was merged.

**Comment by str4d** [comment]
[Force-pushed](https://github.com/zcash/librustzcash/compare/4d17970c51079a323327cf1ae240846e9c05d6e9..306d5a81e0039525fcf0dfdf523b70dc744499bc) to add (interim) changelog entries.

File: zcash_primitives/src/sapling/prover.rs [hunk 1]
```diff
@@ -1,16 +1,75 @@
...
     },
     transaction::components::{Amount, GROTH_PROOF_SIZE},
 };
 
 use super::{Diversifier, PaymentAddress, ProofGenerationKey, Rseed};
 
+/// Interface for creating Sapling Spend proofs.
+pub trait SpendProver {
+    /// The proof type created by this prover.
+    type Proof;
```
Comment (str4d) [inline zcash_primitives/src/sapling/prover.rs:20]: Technically now that `zcash_primitives` depends on `bellman` (after rebasing in #999), we could remove this associated type and just use `Proof<Bls12>`, but to some extent that does depend on the trait implementation (e.g. a remote prover might just return bytes here).

**Comment by str4d** [comment]
[Force-pushed](https://github.com/zcash/librustzcash/compare/306d5a81e0039525fcf0dfdf523b70dc744499bc..0d46fe72cc9e81154e78f41b6936f58a5fd6ccd2) to silence clippy lint.

File: zcash_proofs/CHANGELOG.md [hunk 2]
```diff
@@ -6,6 +6,21 @@ and this library adheres to Rust's notion of
...
+
+### Changed
+- The new `SpendParameters` and `OutputParameters` types are used in the
+  following places:
+  - `zcash_proofs::ZcashParameters::{spend_params, output_params}` fields.
+  - `zcash_proofs::sapling::prover`:
+    - `SaplingProvingContext::{spend_proof, output_proof}` (the `proving_key`
+      arguments).
+- `zcash_proofs::sapling::prover`:
+  - The `verifying_key` argument `SaplingProvingContext::spend_proof` has been
```
Comment (ebfull) [inline zcash_proofs/CHANGELOG.md:20]: nit: typo (should be "argument of")

**Review by ebfull — APPROVED** [review]

File: zcash_proofs/CHANGELOG.md (hunk 2 above)
Comment (str4d) [reply zcash_proofs/CHANGELOG.md:20]: I'll fix that later (possibly by removing the line). If nowhere else, we'll fix this in the next release PR.

File: zcash_proofs/src/prover.rs [hunk 3]
```diff
@@ -13,17 +13,23 @@ use zcash_primitives::{
...
 #[cfg(feature = "local-prover")]
 use crate::{default_params_folder, SAPLING_OUTPUT_NAME, SAPLING_SPEND_NAME};
 
 /// An implementation of [`TxProver`] using Sapling Spend and Output parameters from
 /// locally-accessible paths.
 pub struct LocalTxProver {
-    spend_params: Parameters<Bls12>,
+    spend_params: SpendParameters,
+    // TODO: Either re-introduce verification-after-proving (once the verifier is
+    // refactored), or remove this.
```
Comment (daira) [inline zcash_proofs/src/prover.rs:29]: Filed #1002.

**Review by daira — APPROVED** [review]
utACK

**Pull request merged**